### ISSUE TITLE
Image server: objects returning 500 errors (increased frequency since November 19)

### ISSUE THREAD

**Issue by anarchivist** [sul-dlss/stacks#843]
Reported in [VUF-8555](https://jirasul.stanford.edu/jira/browse/VUF-8555): Thumbnails, as well as any other derivatives, of some objects aren't loading.

This was reported for an atlas ([SW record](https://searchworks.stanford.edu/view/10453652), [PURL (virtual object)](http://purl.stanford.edu/xk347zc4258)). @andrewjbtw noted that the images for the component objects, e.g. [yx039ns9157](https://purl.stanford.edu/yx039ns9157), aren't showing as well. 

Andrew has thus verified the following: 

> 1. Is there a JP2 file on Stacks? Yes.
> 2. Are the permissions for the JP2 file correct? Yes.
> 3. Can you download the JP2 file from the download option in the viewer? Yes.
> 4. Can you download any derivative JPG files from the viewer? No.

This may be an issue based on the image size of the source JP2 (10k x 7k pixels), but it is a bit smaller than other images that we've had issues with previously (e.g. Japanese military maps like [this one](https://purl.stanford.edu/kr585qy0167), which is 14.5kx11.7k pixels).

**Edit:** NR is [also showing](https://onenr.io/0LkjnaDd0Ro) that we're seeing a number of 500 errors (~130 in a 30 minute period between 2:24 pm and 2:54 pm). This might be a broader issue than just this object and looks like it might have [started around November 19](https://onenr.io/01OwveG11Qv).

**Comment by cbeer** [comment]
```
% curl "https://sul-stacks-uat-a.stanford.edu/image/iiif/yx039ns9157/3910a/full/639,/0/default.jpg" -k
500 Internal Server Error

I/O error reading header! (command output: terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
)


java.io.IOException: I/O error reading header! (command output: terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
)
	at edu.illinois.library.cantaloupe.resource.ImageRepresentation.copyOrProcess(ImageRepresentation.java:149)
	at edu.illinois.library.cantaloupe.resource.ImageRepresentation.write(ImageRepresentation.java:127)
	at edu.illinois.library.cantaloupe.resource.iiif.v2.ImageResource.doGET(ImageResource.java:213)
	at edu.illinois.library.cantaloupe.resource.HandlerServlet.handle(HandlerServlet.java:98)
	at edu.illinois.library.cantaloupe.resource.HandlerServlet.doGet(HandlerServlet.java:36)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:622)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:292)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:528)
	at org.apache.coyote.ajp.AbstractAjpProcessor.process(AbstractAjpProcessor.java:873)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:687)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1538)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1494)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
Caused by: edu.illinois.library.cantaloupe.processor.ProcessorException: I/O error reading header! (command output: terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
)
	at edu.illinois.library.cantaloupe.processor.KakaduDemoProcessor.process(KakaduDemoProcessor.java:357)
	at edu.illinois.library.cantaloupe.resource.ImageRepresentation.process(ImageRepresentation.java:180)
	at edu.illinois.library.cantaloupe.resource.ImageRepresentation.copyOrProcess(ImageRepresentation.java:147)
	... 27 more
Caused by: javax.imageio.IIOException: I/O error reading header!
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.readHeader(TIFFImageReader.java:505)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.locateImage(TIFFImageReader.java:643)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.seekToImage(TIFFImageReader.java:773)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.seekToImage(TIFFImageReader.java:760)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.prepareRead(TIFFImageReader.java:1530)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.read(TIFFImageReader.java:1724)
	at javax.imageio.ImageReader.read(ImageReader.java:939)
	at edu.illinois.library.cantaloupe.processor.codec.AbstractIIOImageReader.read(AbstractIIOImageReader.java:339)
	at edu.illinois.library.cantaloupe.processor.KakaduDemoProcessor.processInUnix(KakaduDemoProcessor.java:441)
	at edu.illinois.library.cantaloupe.processor.KakaduDemoProcessor.process(KakaduDemoProcessor.java:341)
	... 29 more
Caused by: java.io.EOFException
	at javax.imageio.stream.ImageInputStreamImpl.readShort(ImageInputStreamImpl.java:229)
	at javax.imageio.stream.ImageInputStreamImpl.readUnsignedShort(ImageInputStreamImpl.java:242)
	at it.geosolutions.imageioimpl.plugins.tiff.TIFFImageReader.readHeader(TIFFImageReader.java:451)
	... 38 more

```

**Comment by cbeer** [comment]
I've traced this down into our old version of kakadu, which is having trouble reading the image. I don't think it's related to the size of the image, but something peculiar about how the JP2000 was created.

The only obvious difference I see between these images and ones that work is that the broken ones have an additional UUIDBox element containing a huge blob of XMP metadata. I'm not sure of the provenance of these images, but it seems similar to other issues we've had with Cantaloupe/Djatoka.

**Comment by andrewjbtw** [comment]
From the workflow history, it looks like the JP2s were created in 2016 with the assemblyWF image processing tools in place at that time.

**Comment by anarchivist** [comment]
Other objects for reference that are exhibiting this behavior:

| druid | source |  note |
| ------ | ------- |  ---- | 
| dz059fy2076 | rumsey | displays nothing  | 
| rw783dp7425 | spec (fitch) |  displays nothing |
| kt299jh2316 | ua | displays nothing |
| yk445px2923 | ua | mirador throws a `TypeError: Failed to Fetch` |

**Comment by cbeer** [comment]
yk445px2923 is a different failure (also previously known and definitely had to do with how the JP2000 was generated)

**Comment by cbeer** [comment]
And, for what it's worth, rw783dp7425  and kt299jh2316 also have that extra UUIDBox with XMP metadata.

**Comment by anarchivist** [comment]
Just to recap our assumptions and work so far:

1) @cbeer's presumption is that the extra UUIDBox containing XMP in the JPEG 2000 file is what's causing Kakadu to give up processing the file.
2) @andrewjbtw has tried reaccessioning `yx039ns9157`, and has confirmed the file is different. The new file, however, still contains the XMP metadata in the JP2.
3) I've looked at the files from yx039ns9157 and rw783dp7425, and the commonality is that they both have XMP UUIDBoxes, although the metadata in each varies. 
4) Our presumption is that the best way to fix this requires intervention and work on the image server. This will not be a short fix and will require a focused work cycle.

**Comment by anarchivist** [comment]
Also, interesting wrinkle re: my item 3 in the last issue. The JP2 files for yx039ns9157 and rw783dp7425 both declare their color profile, to be `sRGB` but their embedded XMP metadata in the uuidBox for both files specifies that they are `Apple RGB`. 

Seems weird, and feels similar, at least loosely, to sul-dlss/common-accessioning#787. However, it feels more likely that Kakadu is having a different issue parsing the uuidBox containing XMP more broadly.

**Comment by anarchivist** [comment]
It turns out the XMP data is a red herring - the issue is how Kakadu 7.10.2 handles the `uuidBox` containing IPTC data. We ran into this previously as sul-dlss/stacks#579. To wit:

```
[lyberadmin@sul-imageserver-uat ~]$ kdu_expand -v
This is Kakadu's "kdu_expand" application.
    Compiled against the Kakadu core system, version v7.10.2
    Current core system version is v7.10.2
[lyberadmin@sul-imageserver-uat ~]$ kdu_expand -i /stacks/yx/039/ns/9157/3910a.jp2 -o /tmp/3910a.tiff
Copying XMP  box info, size =   15466
Copying IPTC box info, size =      -4
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted
[lyberadmin@sul-imageserver-uat ~]$ kdu_expand -i /stacks/rw/783/dp/7425/024_BH_GetOut_2662_32A_edited-1.jp2 -o /tmp/b24.tiff
Copying XMP  box info, size =   16790
Copying IPTC box info, size =      -8
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted
```

@cbeer's comment on that issue still stands:

> Kakadu 8 fixes this particular issue, at least, but is incompatible with our Centos 7 servers (despite their claims to the contrary).

**Comment by andrewjbtw** [comment]
Not sure if this is the same issue, but this recently accessioned item is also showing 500 errors: https://purl.stanford.edu/bb976rq0538

**Comment by anarchivist** [comment]
@andrewjbtw It's the same issue. 

Potential near-term mitigation strategies:

1) Downgrade to Kakadu 7.9, which is what we did to fix #579. This should (probably) be considered a temporary fix, because for some reason the packages were updated back to 7.10.2 between when we rolled them back in April 2020-ish and now. 
2) Track down one of the latest Kakadu 7.10 releases. I believe the last was 7.10.7, and the [release notes](https://kakadusoftware.com/wp-content/uploads/Overview.txt) suggest that maybe this fix was possibly added in 7.10.5. This is probably a "better" temporary fix, but perhaps more work because we have to track down a package/binaries.

Longer term mitigation is most lilely to upgrade Kakadu or replace it with something else (Grok, OpenJPEG, etc. )
